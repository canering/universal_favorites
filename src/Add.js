/*global chrome browser*/
import React from "react";
import PropTypes from "prop-types";
import { browserName } from "react-device-detect";
import urlParse from "url-parse";
import Back from "./images/back.png";

class Add extends React.Component {

    constructor(props){
        super(props);

        this.state = {
            name:"",
            url:"",
            isFolder:false
        };
    }

    componentDidMount = () => {
        if(browserName === "Firefox" || browserName === "Edge") {
            browser.tabs.query({"active": true, "lastFocusedWindow": true}).then((tabs) => {
                let url = tabs[0].url;
                let hostName = urlParse(url, true).hostname;
                hostName = hostName.replace("www.", "");

                this.setState({url:url, name:hostName});
            }).catch(err => console.error(err));
        } else {
            chrome.tabs.query({"active": true, "lastFocusedWindow": true}, (tabs) => {
                let url = tabs[0].url;
                let hostName = urlParse(url, true).hostname;
                hostName = hostName.replace("www.", "");

                this.setState({url:url, name:hostName});
            });
        }
    }

    static get propTypes(){
        return({
            goBack:PropTypes.func,
            addFolder:PropTypes.func,
            addLink:PropTypes.func
        });
    }

    render(){
        return(
            <div className="menu">
                <button id="onlyBack" className="menuButton" onClick={this.backToFavs}>
                    <img src={Back} alt="back" className="icon" />
                </button><br />

                <label id="toggle-bar">
                    <button className={this.state.isFolder ? "toggle-bar-disabled-button" : "toggle-bar-enabled-button"} 
                        disabled={!this.state.isFolder} style={{borderRight:"none", borderRadius:"8px 0 0 8px"}} onClick={() => this.handleChecked(false)}>Link</button>
                    <button className={this.state.isFolder ? "toggle-bar-enabled-button" : "toggle-bar-disabled-button"} 
                        disabled={this.state.isFolder} style={{borderRadius:"0 8px 8px 0"}} onClick={() => this.handleChecked(true)}>Folder</button>
                </label><br />

				Name: <input className="inputField" type="text" onChange={this.updateName} value={this.state.name} /><br />
				URL: <input className="inputField" type="text" onChange={this.updateUrl} value={this.state.url} disabled={this.state.isFolder} /><br />

                <button className="plainButton" onClick={this.submit}>Submit</button>
            </div>
        );
    }

    handleChecked = (isFolder) => {
        this.setState({isFolder:isFolder});
    }
    updateName = (event) => {
        this.setState({name:event.target.value});
    }
    updateUrl = (event) => {
        this.setState({url:event.target.value});
    }
    backToFavs = () => {
        this.props.goBack();
    }
    submit = () => {
        if(this.state.name === "") {
            return;
        }

        if(this.state.isFolder){
            this.props.addFolder(this.state.name);
        }
        else{
            if(this.state.url === "") {
                return;
            }

            this.props.addLink(this.state.name, this.state.url);
        }
    }
}

export default Add;