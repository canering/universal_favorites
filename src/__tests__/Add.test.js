import React from "react";
import {shallow, configure} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import renderer from "react-test-renderer";
import Add from "../Add";
import chrome from "sinon-chrome";

window.chrome = chrome;
chrome.tabs.query.yields([{url:"http://example.com"}]);

configure({adapter: new Adapter()});

const props = {
    goBack: jest.fn(),
    addFolder: jest.fn(),
    addLink: jest.fn()
};

afterEach(() => {
    jest.clearAllMocks();
});

test("Add Menu Snapshot Test", () => {

    const component = renderer.create(<Add {...props} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    component.root.findByProps({className:"toggle-bar-disabled-button"}).props.onClick();

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("when back is clicked than goBack should be called", () => {
    const component = shallow(<Add {...props} />);
    component.find("#onlyBack").prop("onClick")();

    expect(props.goBack).toHaveBeenCalledTimes(1);
});

test("when link is submitted then addLink should be called", () => {
    const component = shallow(<Add {...props} />);
    component.find(".plainButton").prop("onClick")();

    expect(props.addLink).toHaveBeenCalledTimes(1);
    expect(props.addLink).toHaveBeenCalledWith("example.com", "http://example.com");
});

test("when folder is submitted then addFolder should be called", () => {
    const component = shallow(<Add {...props} />);
    component.find(".toggle-bar-disabled-button").prop("onClick")();
    component.find(".plainButton").prop("onClick")();

    expect(props.addFolder).toHaveBeenCalledTimes(1);
    expect(props.addFolder).toHaveBeenCalledWith("example.com");
});