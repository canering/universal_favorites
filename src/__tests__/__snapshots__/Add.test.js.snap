// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Menu Snapshot Test 1`] = `
<div
  className="menu"
>
  <button
    className="menuButton"
    id="onlyBack"
    onClick={[Function]}
  >
    <img
      alt="back"
      className="icon"
      src="back.png"
    />
  </button>
  <br />
  <label
    id="toggle-bar"
  >
    <button
      className="toggle-bar-enabled-button"
      disabled={true}
      onClick={[Function]}
      style={
        Object {
          "borderRadius": "8px 0 0 8px",
          "borderRight": "none",
        }
      }
    >
      Link
    </button>
    <button
      className="toggle-bar-disabled-button"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "borderRadius": "0 8px 8px 0",
        }
      }
    >
      Folder
    </button>
  </label>
  <br />
  Name: 
  <input
    className="inputField"
    onChange={[Function]}
    type="text"
    value="example.com"
  />
  <br />
  URL: 
  <input
    className="inputField"
    disabled={false}
    onChange={[Function]}
    type="text"
    value="http://example.com"
  />
  <br />
  <button
    className="plainButton"
    onClick={[Function]}
  >
    Submit
  </button>
</div>
`;

exports[`Add Menu Snapshot Test 2`] = `
<div
  className="menu"
>
  <button
    className="menuButton"
    id="onlyBack"
    onClick={[Function]}
  >
    <img
      alt="back"
      className="icon"
      src="back.png"
    />
  </button>
  <br />
  <label
    id="toggle-bar"
  >
    <button
      className="toggle-bar-disabled-button"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "borderRadius": "8px 0 0 8px",
          "borderRight": "none",
        }
      }
    >
      Link
    </button>
    <button
      className="toggle-bar-enabled-button"
      disabled={true}
      onClick={[Function]}
      style={
        Object {
          "borderRadius": "0 8px 8px 0",
        }
      }
    >
      Folder
    </button>
  </label>
  <br />
  Name: 
  <input
    className="inputField"
    onChange={[Function]}
    type="text"
    value="example.com"
  />
  <br />
  URL: 
  <input
    className="inputField"
    disabled={true}
    onChange={[Function]}
    type="text"
    value="http://example.com"
  />
  <br />
  <button
    className="plainButton"
    onClick={[Function]}
  >
    Submit
  </button>
</div>
`;
